File: cli/internal/validation/shared.go
package validation

import (
	"fmt"
	"path"
	"strings"
)

// CheckArchiveSize validates the archive size against a maximum.
func CheckArchiveSize(data []byte, maxBytes int) *ValidationError {
	if len(data) > maxBytes {
		return &ValidationError{
			Code:    "SIZE_EXCEEDED",
			Message: fmt.Sprintf("Archive size %s exceeds maximum %s", formatBytes(len(data)), formatBytes(maxBytes)),
		}
	}
	return nil
}

// CheckPathTraversal checks for path traversal attacks in archive file paths.
func CheckPathTraversal(files map[string][]byte) *ValidationError {
	for filePath := range files {
		if strings.Contains(filePath, "..") || path.IsAbs(filePath) {
			return &ValidationError{
				Code:    "PATH_TRAVERSAL",
				Message: fmt.Sprintf("Dangerous path detected: %s", filePath),
				Path:    filePath,
			}
		}
	}
	return nil
}

// ReadFileFromArchive reads a file from the archive, checking both direct path and under rootDir.
func ReadFileFromArchive(contents *ArchiveContents, relativePath string) string {
	if data, ok := contents.Files[relativePath]; ok {
		return string(data)
	}
	if contents.RootDir != "" {
		if data, ok := contents.Files[contents.RootDir+"/"+relativePath]; ok {
			return string(data)
		}
	}
	return ""
}

// FileExistsInArchive checks whether a file exists in the archive.
func FileExistsInArchive(contents *ArchiveContents, relativePath string) bool {
	if _, ok := contents.Files[relativePath]; ok {
		return true
	}
	if contents.RootDir != "" {
		if _, ok := contents.Files[contents.RootDir+"/"+relativePath]; ok {
			return true
		}
	}
	return false
}

// DirExistsInArchive checks whether a directory (prefix) exists in the archive.
func DirExistsInArchive(contents *ArchiveContents, dirPath string) bool {
	normalizedDir := dirPath
	if !strings.HasSuffix(normalizedDir, "/") {
		normalizedDir += "/"
	}
	for filePath := range contents.Files {
		if strings.HasPrefix(filePath, normalizedDir) {
			return true
		}
		if contents.RootDir != "" && strings.HasPrefix(filePath, contents.RootDir+"/"+normalizedDir) {
			return true
		}
	}
	return false
}

// GetFilesInDir returns all file paths under a given directory in the archive.
func GetFilesInDir(contents *ArchiveContents, dirPath string) []string {
	normalizedDir := dirPath
	if !strings.HasSuffix(normalizedDir, "/") {
		normalizedDir += "/"
	}
	var result []string
	for filePath := range contents.Files {
		if strings.HasPrefix(filePath, normalizedDir) {
			result = append(result, filePath)
		} else if contents.RootDir != "" && strings.HasPrefix(filePath, contents.RootDir+"/"+normalizedDir) {
			result = append(result, filePath[len(contents.RootDir)+1:])
		}
	}
	return result
}

func formatBytes(bytes int) string {
	if bytes < 1024 {
		return fmt.Sprintf("%dB", bytes)
	}
	if bytes < 1024*1024 {
		return fmt.Sprintf("%.1fKB", float64(bytes)/1024)
	}
	return fmt.Sprintf("%.1fMB", float64(bytes)/(1024*1024))
}

File: packages/validation/src/shared.ts
import * as tar from 'tar';
import AdmZip from 'adm-zip';
import path from 'path';
import { tmpdir } from 'os';
import { mkdtemp, rm } from 'fs/promises';
import { readdirSync, readFileSync } from 'fs';
import type { ArchiveContents } from './types.js';
import type { ValidationError } from '@lootprotocol/shared-types';

export async function extractArchive(
  buffer: Buffer,
  filename: string,
): Promise<ArchiveContents> {
  if (filename.endsWith('.zip')) {
    return extractZip(buffer);
  } else if (filename.endsWith('.tar.gz') || filename.endsWith('.tgz')) {
    return extractTarGz(buffer);
  }
  throw new Error(`Unsupported archive format: ${filename}`);
}

function extractZip(buffer: Buffer): ArchiveContents {
  const zip = new AdmZip(buffer);
  const files = new Map<string, Buffer>();
  for (const entry of zip.getEntries()) {
    if (!entry.isDirectory) {
      const normalized = normalizePath(entry.entryName);
      files.set(normalized, entry.getData());
    }
  }
  return { files, rootDir: detectRootDir(files) };
}

async function extractTarGz(buffer: Buffer): Promise<ArchiveContents> {
  const tmpDir = await mkdtemp(path.join(tmpdir(), 'lootprotocol-'));
  try {
    const extractor = tar.extract({ cwd: tmpDir, gzip: true });
    await new Promise<void>((resolve, reject) => {
      extractor.on('end', resolve);
      extractor.on('error', reject);
      extractor.end(buffer);
    });
    const files = readDirRecursive(tmpDir, tmpDir);
    return { files, rootDir: detectRootDir(files) };
  } finally {
    await rm(tmpDir, { recursive: true, force: true });
  }
}

function readDirRecursive(dir: string, baseDir: string): Map<string, Buffer> {
  const files = new Map<string, Buffer>();
  const entries = readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      const subFiles = readDirRecursive(fullPath, baseDir);
      for (const [k, v] of subFiles) {
        files.set(k, v);
      }
    } else {
      const relativePath = normalizePath(path.relative(baseDir, fullPath));
      const content = readFileSync(fullPath);
      files.set(relativePath, content);
    }
  }
  return files;
}

function normalizePath(filePath: string): string {
  return filePath.replace(/\\/g, '/').replace(/^\/+/, '');
}

function detectRootDir(files: Map<string, Buffer>): string {
  const paths = Array.from(files.keys());
  if (paths.length === 0) return '';

  const firstSegments = new Set<string>();
  for (const p of paths) {
    const parts = p.split('/');
    if (parts.length > 1) {
      firstSegments.add(parts[0]);
    }
  }

  if (firstSegments.size === 1) {
    const rootDir = Array.from(firstSegments)[0];
    const allUnderRoot = paths.every((p) => p.startsWith(`${rootDir}/`));
    if (allUnderRoot) return rootDir;
  }

  return '';
}

export function checkArchiveSize(
  buffer: Buffer,
  maxBytes: number,
): ValidationError | null {
  if (buffer.length > maxBytes) {
    return {
      code: 'SIZE_EXCEEDED',
      message: `Archive size ${formatBytes(buffer.length)} exceeds maximum ${formatBytes(maxBytes)}`,
    };
  }
  return null;
}

export function checkPathTraversal(
  files: Map<string, Buffer>,
): ValidationError | null {
  for (const filePath of files.keys()) {
    if (filePath.includes('..') || path.isAbsolute(filePath)) {
      return {
        code: 'PATH_TRAVERSAL',
        message: `Dangerous path detected: ${filePath}`,
        path: filePath,
      };
    }
  }
  return null;
}

export function readFileFromArchive(
  contents: ArchiveContents,
  relativePath: string,
): string | null {
  const buffer =
    contents.files.get(relativePath) ||
    (contents.rootDir
      ? contents.files.get(`${contents.rootDir}/${relativePath}`)
      : undefined);
  return buffer ? buffer.toString('utf-8') : null;
}

export function fileExistsInArchive(
  contents: ArchiveContents,
  relativePath: string,
): boolean {
  return (
    contents.files.has(relativePath) ||
    (!!contents.rootDir &&
      contents.files.has(`${contents.rootDir}/${relativePath}`))
  );
}

export function dirExistsInArchive(
  contents: ArchiveContents,
  dirPath: string,
): boolean {
  const normalizedDir = dirPath.endsWith('/') ? dirPath : `${dirPath}/`;
  for (const filePath of contents.files.keys()) {
    if (filePath.startsWith(normalizedDir)) return true;
    if (
      contents.rootDir &&
      filePath.startsWith(`${contents.rootDir}/${normalizedDir}`)
    ) {
      return true;
    }
  }
  return false;
}

export function getFilesInDir(
  contents: ArchiveContents,
  dirPath: string,
): string[] {
  const normalizedDir = dirPath.endsWith('/') ? dirPath : `${dirPath}/`;
  const result: string[] = [];
  for (const filePath of contents.files.keys()) {
    if (filePath.startsWith(normalizedDir)) {
      result.push(filePath);
    } else if (
      contents.rootDir &&
      filePath.startsWith(`${contents.rootDir}/${normalizedDir}`)
    ) {
      result.push(filePath.slice(contents.rootDir.length + 1));
    }
  }
  return result;
}

function formatBytes(bytes: number): string {
  if (bytes < 1024) return `${bytes}B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)}KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)}MB`;
}
